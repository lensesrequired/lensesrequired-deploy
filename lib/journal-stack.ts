import * as cdk from 'aws-cdk-lib';
import {SecretValue} from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import {ProjectionType} from 'aws-cdk-lib/aws-dynamodb';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';

export class JournalStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.TableV2(this, 'JournalTable', {
      tableName: 'journal-table',
      billing: dynamodb.Billing.onDemand({
        maxReadRequestUnits: 50,
        maxWriteRequestUnits: 25
      }),
      partitionKey: { name: 'PK', type: dynamodb.AttributeType.STRING },
      sortKey: { name: 'SK', type: dynamodb.AttributeType.STRING },
      deletionProtection: true,
      timeToLiveAttribute: 'ttl',
      removalPolicy: cdk.RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      pointInTimeRecoverySpecification: {
        pointInTimeRecoveryEnabled: true,
      }
    });

    const execUser = new iam.User(this, 'JournalUser', {
      userName: 'journal-exec-user'
    })
    execUser.attachInlinePolicy(new iam.Policy(this, 'JournalExecPolicy', {
      statements: [
        new iam.PolicyStatement({
          sid: 'DynamoDBReadAccess',
          actions: [
            "dynamodb:List*",
            "dynamodb:DescribeReservedCapacity*",
            "dynamodb:DescribeLimits",
            "dynamodb:DescribeTimeToLive"
          ],
          resources: ['*']
        }),
        new iam.PolicyStatement({
          sid: 'DynamoDBWriteAccess',
          actions: [
            "dynamodb:BatchGet*",
            "dynamodb:DescribeStream",
            "dynamodb:DescribeTable",
            "dynamodb:Get*",
            "dynamodb:Query",
            "dynamodb:Scan",
            "dynamodb:BatchWrite*",
            "dynamodb:CreateTable",
            "dynamodb:Delete*",
            "dynamodb:Update*",
            "dynamodb:PutItem"
          ],
          resources: [table.tableArn, `${table.tableArn}/*`]
        })
      ]
    }))
    execUser.node.addDependency(table)

    const accessKey = new iam.AccessKey(this, 'JournalExecUserAccessKey', { user: execUser });
    accessKey.node.addDependency(execUser);

    const secret = new secretsmanager.Secret(this, 'JournalAccessKeySecret', {
      secretObjectValue: {
        Username: SecretValue.unsafePlainText(execUser.userName),
        AccessKeyId: SecretValue.unsafePlainText(accessKey.accessKeyId),
        SecretAccessKey: accessKey.secretAccessKey
      },
    });
    secret.node.addDependency(accessKey)
  }
}
