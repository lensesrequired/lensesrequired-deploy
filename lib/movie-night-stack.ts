import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as iam from 'aws-cdk-lib/aws-iam';

export class MovieNightStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.TableV2(this, 'MovieNightTable', {
      tableName: 'movie-night-table',
      billing: dynamodb.Billing.onDemand({
        maxReadRequestUnits: 50,
        maxWriteRequestUnits: 25
      }),
      partitionKey: { name: 'PK', type: dynamodb.AttributeType.STRING },
      sortKey: { name: 'SK', type: dynamodb.AttributeType.STRING },
      deletionProtection: true,
      timeToLiveAttribute: 'ttl',
      removalPolicy: cdk.RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      pointInTimeRecoverySpecification: {
        pointInTimeRecoveryEnabled: true,
      },
      globalSecondaryIndexes: [
        {
          indexName: 'GSI1',
          partitionKey: { name: 'SK', type: dynamodb.AttributeType.STRING },
          sortKey: { name: 'GSI_SK', type: dynamodb.AttributeType.STRING },
        },
      ]
    });

    const execUser = new iam.User(this, 'MovieNightUser', {
      userName: 'movie-night-exec-user'
    })
    execUser.attachInlinePolicy(new iam.Policy(this, 'MovieNightExecPolicy', {
      statements: [
        new iam.PolicyStatement({
          sid: 'DynamoDBReadAccess',
          actions: [
            "dynamodb:List*",
            "dynamodb:DescribeReservedCapacity*",
            "dynamodb:DescribeLimits",
            "dynamodb:DescribeTimeToLive"
          ],
          resources: ['*']
        }),
        new iam.PolicyStatement({
          sid: 'DynamoDBWriteAccess',
          actions: [
            "dynamodb:BatchGet*",
            "dynamodb:DescribeStream",
            "dynamodb:DescribeTable",
            "dynamodb:Get*",
            "dynamodb:Query",
            "dynamodb:Scan",
            "dynamodb:BatchWrite*",
            "dynamodb:CreateTable",
            "dynamodb:Delete*",
            "dynamodb:Update*",
            "dynamodb:PutItem"
          ],
          resources: [table.tableArn]
        })
      ]
    }))
    execUser.node.addDependency(table)
  }
}
