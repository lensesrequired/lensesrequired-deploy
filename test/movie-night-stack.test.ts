import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {MovieNightStack} from '../lib/movie-night-stack';

describe('MovieNightStack', () => {
  it('matches snapshot', () => {
    const app = new cdk.App({context: {account: 'test'}});
    const stack = new MovieNightStack(app, 'MyTestStack');

    const template = Template.fromStack(stack)
    expect(template.toJSON()).toMatchSnapshot()
  })
});
