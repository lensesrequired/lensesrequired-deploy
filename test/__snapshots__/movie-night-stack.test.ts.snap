// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieNightStack matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MovieNightExecPolicy39E72BA7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:List*",
                "dynamodb:DescribeReservedCapacity*",
                "dynamodb:DescribeLimits",
                "dynamodb:DescribeTimeToLive",
              ],
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "DynamoDBReadAccess",
            },
            {
              "Action": [
                "dynamodb:BatchGet*",
                "dynamodb:DescribeStream",
                "dynamodb:DescribeTable",
                "dynamodb:Get*",
                "dynamodb:Query",
                "dynamodb:Scan",
                "dynamodb:BatchWrite*",
                "dynamodb:CreateTable",
                "dynamodb:Delete*",
                "dynamodb:Update*",
                "dynamodb:PutItem",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MovieNightTableF882A7F3",
                  "Arn",
                ],
              },
              "Sid": "DynamoDBWriteAccess",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MovieNightExecPolicy39E72BA7",
        "Users": [
          {
            "Ref": "MovieNightUserD03138EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MovieNightTableF882A7F3": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI_SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "GSI1",
            "KeySchema": [
              {
                "AttributeName": "SK",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "GSI_SK",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "DeletionProtectionEnabled": true,
            "GlobalSecondaryIndexes": [
              {
                "IndexName": "GSI1",
              },
            ],
            "PointInTimeRecoverySpecification": {
              "PointInTimeRecoveryEnabled": true,
            },
            "ReadOnDemandThroughputSettings": {
              "MaxReadRequestUnits": 50,
            },
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
        "TableName": "movie-night-table",
        "TimeToLiveSpecification": {
          "AttributeName": "ttl",
          "Enabled": true,
        },
        "WriteOnDemandThroughputSettings": {
          "MaxWriteRequestUnits": 25,
        },
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
    "MovieNightUserD03138EC": {
      "DependsOn": [
        "MovieNightTableF882A7F3",
      ],
      "Properties": {
        "UserName": "movie-night-exec-user",
      },
      "Type": "AWS::IAM::User",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
